server/service/async/async.go

```diff
@@ -17,94 +18,75 @@ import (
+	pool        fleet.RedisPool
+	clock       clock.Clock
+	taskConfigs map[string]config.AsyncProcessingConfig
 	seenHostSet seenHostSet
 }
 
+// NewTask configures and returns a Task.
+func NewTask(ds fleet.Datastore, pool fleet.RedisPool, clck clock.Clock, conf config.OsqueryConfig) *Task {
+	taskCfgs := make(map[string]config.AsyncProcessingConfig)
+	taskCfgs[config.AsyncTaskLabelMembership] = conf.AsyncConfigForTask(config.AsyncTaskLabelMembership)
+	taskCfgs[config.AsyncTaskPolicyMembership] = conf.AsyncConfigForTask(config.AsyncTaskPolicyMembership)
+	taskCfgs[config.AsyncTaskHostLastSeen] = conf.AsyncConfigForTask(config.AsyncTaskHostLastSeen)
+	return &Task{
+		datastore:   ds,
+		pool:        pool,
+		clock:       clck,
+		taskConfigs: taskCfgs,
+	}
+}
+
 // Collect runs the various collectors as distinct background goroutines if
 // async processing is enabled.  Each collector will stop processing when ctx
 // is done.
-func (t *Task) StartCollectors(ctx context.Context, jitterPct int, logger kitlog.Logger) {
-	if !t.AsyncEnabled {
-		level.Debug(logger).Log("task", "async disabled, not starting collectors")
-		return
-	}
-	level.Debug(logger).Log("task", "async enabled, starting collectors", "interval", t.CollectorInterval, "jitter", jitterPct)
-
+func (t *Task) StartCollectors(ctx context.Context, logger kitlog.Logger) {
 	collectorErrHandler := func(name string, err error) {
 		level.Error(logger).Log("err", fmt.Sprintf("%s collector", name), "details", err)
 		sentry.CaptureException(err)
 	}
 
-	labelColl := &collector{
-		name:         "collect_labels",
-		pool:         t.Pool,
-		ds:           t.Datastore,
-		execInterval: t.CollectorInterval,
-		jitterPct:    jitterPct,
-		lockTimeout:  t.LockTimeout,
-		handler:      t.collectLabelQueryExecutions,
-		errHandler:   collectorErrHandler,
-	}
-
-	policyColl := &collector{
-		name:         "collect_policies",
-		pool:         t.Pool,
-		ds:           t.Datastore,
-		execInterval: t.CollectorInterval,
-		jitterPct:    jitterPct,
-		lockTimeout:  t.LockTimeout,
-		handler:      t.collectPolicyQueryExecutions,
-		errHandler:   collectorErrHandler,
-	}
-
-	lastSeenColl := &collector{
-		name:         "collect_last_seen",
-		pool:         t.Pool,
-		ds:           t.Datastore,
-		execInterval: t.CollectorInterval,
-		jitterPct:    jitterPct,
-		lockTimeout:  t.LockTimeout,
-		handler:      t.collectHostsLastSeen,
-		errHandler:   collectorErrHandler,
+	handlers := map[string]collectorHandlerFunc{
+		config.AsyncTaskLabelMembership:  t.collectLabelQueryExecutions,
+		config.AsyncTaskPolicyMembership: t.collectPolicyQueryExecutions,
+		config.AsyncTaskHostLastSeen:     t.collectHostsLastSeen,
 	}
+	for task, cfg := range t.taskConfigs {
+		if !cfg.Enabled {
+			level.Debug(logger).Log("task", "async disabled, not starting collector", "name", task)
+			continue
+		}
 
-	colls := []*collector{labelColl, policyColl, lastSeenColl}
-	for _, coll := range colls {
+		cfg := cfg
```

Comment:
is this intended?

Reply:
Ah yeah, that's a Go gotcha, there's a goroutine started later in the loop with a function that captures the `cfg` variable, if you use the iteration variable it will get the value of the last iteration, regardless of what the value was when the goroutine started (that has to do with function closures, not with goroutines specifically, e.g.: https://go.dev/play/p/FCHJmFxnB7y). That's because the iteration variable is the same for all iterations, it just gets assigned new values, so when you capture it, your capture (which is like a pointer to that variable) gets updated too. Using a local variable instead (that shadows the iteration one - not required per se but it's a common idiom to make sure you don't accidentally capture the iteration one by mistake) ensures that you get a fresh copy on each iteration.

I'll add a comment next to that assignment.

Reply:
ohh, I totally missed that. Thank you very much for the explanation!